### DISCUSSION TITLE
Re-Purpose Miss_hit Help

### DISCUSSION THREAD

**Discussion by ThePerfectComputer** [florianschanda/miss_hit#277] [Q&A]
Making this question public here so that more people can see it:

====

Thank you for work excellent work on miss_hit. Although this isn't an externally exposed functionality, I'm wondering if I can use the miss_hit code to build something that takes a matlab source file and spits out an abstract syntax tree.

I imagine there’s already a function in the code that does exactly that?

**Comment by ThePerfectComputer** [comment]
Also, I'm the ``mh_debug_parser`` works well. The code is easy enough to read. I'm wondering if the Abstract Syntax Tree captures comments?

**Reply by florianschanda** [reply]
It does! Specifically for your problem you might also like to know that the docstrings are captured specifically:

https://github.com/florianschanda/miss_hit/blob/f01fbf9cd43c52fdb8fe6dce162c9772637ec199/miss_hit_core/m_ast.py#L329C7-L329C7

To get this to work you need to borrow the post-processing code from mh_style:

https://github.com/florianschanda/miss_hit/blob/f01fbf9cd43c52fdb8fe6dce162c9772637ec199/miss_hit_core/m_parse_utils.py#L31

You will also need to use the special lexer that preserves all tokens in a list, instead of throwing them away after generation.

https://github.com/florianschanda/miss_hit/blob/f01fbf9cd43c52fdb8fe6dce162c9772637ec199/miss_hit_core/mh_style.py#L1068
https://github.com/florianschanda/miss_hit/blob/f01fbf9cd43c52fdb8fe6dce162c9772637ec199/miss_hit_core/mh_style.py#L1123

**Comment by ThePerfectComputer** [comment]
I will note I'm seeing support for comments as tokens.

**Reply by florianschanda** [reply]
Yes, the lexer returns all comments and the parser then throws them away.

In order to associate comments (specifically the docstrings) to functions, classes, or files you need to do some post-processing similar to what mh_style or mh_copyright does:

https://github.com/florianschanda/miss_hit/blob/f01fbf9cd43c52fdb8fe6dce162c9772637ec199/miss_hit/mh_copyright.py#L152-L159

**Comment by ThePerfectComputer** [comment]
How can I make a modification to the matlab AST, and then write it back out to a file?

**Reply by florianschanda** [reply]
This cannot be done, mainly because e.g. comments are NOT part of the AST.
What `mh_style` does is a bit more subtle, we remember the original token stream (including comments) and then dump it back, modified in small ways.
* https://github.com/florianschanda/miss_hit/blob/f01fbf9cd43c52fdb8fe6dce162c9772637ec199/miss_hit_core/m_lexer.py#L1327 (where we dump it back)
* https://github.com/florianschanda/miss_hit/blob/f01fbf9cd43c52fdb8fe6dce162c9772637ec199/miss_hit_core/m_ast.py#L42 (the datastructure used to modify the replay behaviour

If you want to add e.g. comments then you would need to extend the autofix thing to say "add a comment after this line" or similar and modify the replay function to do that.

**Comment by ThePerfectComputer** [comment]
I have some some identifier nodes and am wondering what is a good way to determine the line-col location of the indentifier node?

**Reply by ThePerfectComputer** [reply]
Right now I'm able to call ``.loc()`` on an identifier node, but I'm currently only seeing the line number and no column information.

**Reply by ThePerfectComputer** [reply]
nvm - the following work:
 * ``node.loc().line``
 * ``node.loc().start_col``
 * ``node.loc().end_col``

**Reply by florianschanda** [reply]
Glad you worked it out :)

**Comment by ThePerfectComputer** [comment]
And on to the next question - heh...

If I install miss_hit via ``pip3 install miss-hit-core``, how do I then do something like ``from miss_hit_core.config import Config`` for example?

**Reply by florianschanda** [reply]
I'm not sure I understand that question. That should work as far as I know?

**Reply by ThePerfectComputer** [reply]
It works now. Turns out there was an issue with my pip configuration.

[a comment by an automated account is not shown]
